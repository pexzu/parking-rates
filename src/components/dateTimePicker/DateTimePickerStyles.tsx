import styled from '@emotion/styled';

export const DateTimePickerContainer = styled.div`
  label: DateTimePickerContainer;
`;

export const DateTimePickerInputLabel = styled.div`
  label: DateTimePickerInputLabel;
`;
