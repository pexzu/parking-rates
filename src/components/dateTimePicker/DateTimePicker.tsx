import * as React from 'react';
import DatePicker from 'react-datepicker';
import 'react-datepicker/dist/react-datepicker.css';
import * as E from './DateTimePickerStyles';

export interface IDateTimePickerProps {
  onChange: (date: Date) => void;
  startDate?: Date;
  minDate?: Date;
  maxDate?: Date;
  placeholder: string;
  inputLabel: string;
}

export interface IDateTimePickerState {
  currentMinTime: Date;
  currentMaxTime: Date;
}

export default class DateTimePicker extends React.Component<
  IDateTimePickerProps,
  IDateTimePickerState
> {
  constructor(props: IDateTimePickerProps) {
    super(props);
    this.state = {
      currentMinTime: new Date(new Date(new Date().setMinutes(0)).setHours(0)),
      currentMaxTime: new Date(new Date(new Date().setMinutes(59)).setHours(23)),
    };
  }

  onChange = (currentDate: Date) => {
    currentDate && this.props.onChange(currentDate);
    this.setMinTimeAndMaxTime(currentDate);
  };

  /*This method is done to avoid the inconsistancy with the datepicked time picked component. 
  Once they publish next version we can exclude timebounds as a part of datebounds*/

  setMinTimeAndMaxTime = (date: Date) => {
    const { minDate, maxDate, startDate } = this.props;
    let currentDate = date;
    console.log(startDate && startDate, date);
    if (startDate && date < startDate) {
      debugger;
      currentDate = startDate;
    }
    if (currentDate && minDate && currentDate.getDate() === minDate.getDate()) {
      this.setState({
        currentMinTime: minDate,
      });
    } else {
      this.setState({
        currentMinTime: new Date(new Date(new Date().setMinutes(0)).setHours(0)),
      });
    }

    if (currentDate && maxDate && currentDate.getDate() === maxDate.getDate()) {
      this.setState({
        currentMaxTime: maxDate,
      });
    } else {
      this.setState({
        currentMaxTime: new Date(new Date(new Date().setMinutes(59)).setHours(23)),
      });
    }
  };

  onClick = () => {
    const { minDate, maxDate, startDate } = this.props;
    let currentDate;
    if (startDate) {
      currentDate = startDate;
    } else if (minDate) {
      currentDate = minDate;
    } else if (maxDate) {
      currentDate = maxDate;
    }

    currentDate && this.setMinTimeAndMaxTime(currentDate);
  };

  public render() {
    const { inputLabel, startDate, minDate, maxDate, placeholder } = this.props;
    const { currentMaxTime, currentMinTime } = this.state;
    return (
      <E.DateTimePickerContainer>
        <E.DateTimePickerInputLabel>{inputLabel}</E.DateTimePickerInputLabel>
        <DatePicker
          placeholderText={placeholder}
          selected={startDate}
          onChange={currentDate => currentDate && this.onChange(currentDate)}
          showTimeSelect
          timeFormat='HH:mm'
          timeIntervals={15}
          dateFormat='MMMM d, yyyy h:mm aa'
          timeCaption='time'
          minDate={minDate && minDate}
          maxDate={maxDate && maxDate}
          minTime={currentMinTime}
          maxTime={currentMaxTime}
          onInputClick={() => this.onClick()}
        />
      </E.DateTimePickerContainer>
    );
  }
}
