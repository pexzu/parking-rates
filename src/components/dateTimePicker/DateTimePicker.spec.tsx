import React from 'react';
import Enzyme, { shallow, mount } from 'enzyme';
import DateTimePicker from './DateTimePicker';
import { DateTimePickerContainer, DateTimePickerInputLabel } from './DateTimePickerStyles';
import Adapter from 'enzyme-adapter-react-16';
import renderer from 'react-test-renderer';

Enzyme.configure({ adapter: new Adapter() });

describe('DateTimePicker component', () => {
  test('render', () => {
    const wrapper = shallow(
      <DateTimePicker onChange={() => {}} inputLabel='testLabel' placeholder='placeholder' />
    );
    expect(wrapper.exists()).toBe(true);
  });
});

//Snapshot Testing
it('DateTimePickerContainer renders correctly', () => {
  const tree = renderer.create(<DateTimePickerContainer />).toJSON();
  expect(tree).toMatchSnapshot();
});

it('DateTimePickerInputLabel renders correctly', () => {
  const tree = renderer.create(<DateTimePickerInputLabel />).toJSON();
  expect(tree).toMatchSnapshot();
});
