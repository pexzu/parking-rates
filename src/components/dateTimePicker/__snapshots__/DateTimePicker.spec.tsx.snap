// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DateTimePickerContainer renders correctly 1`] = `
<div
  className="css-mir5xb-DateTimePickerContainer"
/>
`;

exports[`DateTimePickerInputLabel renders correctly 1`] = `
<div
  className="css-ci9z6j-DateTimePickerInputLabel"
/>
`;
